libsolidity/analysis/OverflowChecker.cpp

```diff
@@ -0,0 +1,220 @@
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * @author Rhett <roadriverrail@gmail.com>
+ * @date 2017
+ * Overflow analyzer and checker.
+ */
+
+#include <libsolidity/analysis/OverflowChecker.h>
+#include <libsolidity/analysis/OverflowBounds.h>
+#include <libsolidity/analysis/IntegerOverflowBounds.h>
+#include <memory>
+#include <boost/range/adaptor/reversed.hpp>
+#include <libsolidity/ast/AST.h>
+#include <libsolidity/inlineasm/AsmAnalysis.h>
+#include <libsolidity/inlineasm/AsmAnalysisInfo.h>
+#include <libsolidity/inlineasm/AsmData.h>
+#include <libsolidity/interface/ErrorReporter.h>
+#include <iostream>
+#include <typeinfo>
+using namespace std;
+using namespace dev;
+using namespace dev::solidity;
+using namespace boost::multiprecision;
+
+bool OverflowChecker::checkOverflow(ASTNode const& _contract)
+{
+	try
+	{
+		_contract.accept(*this);
+	}
+	catch (FatalError const&)
+	{
+		// We got a fatal error which required to stop further checking, but we can
+		// continue normally from here.
+		if (m_errorReporter.errors().empty())
+			throw; // Something is weird here, rather throw again.
+	}
+	return Error::containsOnlyWarnings(m_errorReporter.errors());
+}
+
+bool OverflowChecker::visit(Assignment const& _assignment)
+{
+	_assignment.rightHandSide().accept(*this);
+	auto const identifier = dynamic_cast<Identifier const*>(&_assignment.leftHandSide());
+	if (identifier)
+	{
+		shared_ptr<OverflowBounds>bounds = m_overflowInfo[&_assignment.rightHandSide()];
+		auto var = dynamic_cast<VariableDeclaration const*>(identifier->annotation().referencedDeclaration);
+		m_varOverflowInfo[var] = bounds;
+		if (!bounds->fitsType(*(var->annotation().type)))
+			m_errorReporter.warning(_assignment.location(), "Variable assignment may cause overflow.  " + bounds->toString());
+		else
+			m_errorReporter.warning(_assignment.location(), "Variable assignment ok.  Range: "+bounds->toString()+" Storage: "+var->annotation().type->toString(false));
+	}
+
+	return false;
+}
+
+bool OverflowChecker::lockConstantExpression(Expression const& _expression)
+{
+	auto const* rationalNumberType = dynamic_cast<RationalNumberType const*>(_expression.annotation().type.get());
+	if (rationalNumberType)
+	{
+		shared_ptr<IntegerOverflowBounds> bounds = make_shared<IntegerOverflowBounds>(*rationalNumberType);
+		m_overflowInfo[&_expression] = bounds;
+		return true;
+	}
+	return false;
+}
+
+bool OverflowChecker::visit(Literal const& _literal)
+{
+	lockConstantExpression(_literal);
+	return true;
+}
+
+bool OverflowChecker::visit(BinaryOperation const& _operation)
+{
+	if (lockConstantExpression(_operation))
+		return false; // Do not descend
+	else
+	{
+		// Need to do more work...
+		auto const* integerType = dynamic_cast<IntegerType const*>(_operation.annotation().commonType.get());
+		if (integerType)
+		{
+			//First, descend and resolve both sides of the operation.
```

Comment:
Right now, my code can't support function calls.  When I do get to that point, I will definitely *have* to evaluate the expressions in the parameters to a call because a call like `f(x++)` will generate a side-effect on `x`.  So, I will definitely descend all function calls.  In your example, when that descent reaches `a+1`, I wouldn't expect `lockConstantExpression( )` to filter it out because the expression `a+1` can't be a RationalNumberType.